packages/react-router/lib/router/define-route.ts

```diff
@@ -75,13 +73,13 @@ type _LoaderData<
   ClientLoaderHydrate extends boolean,
   HydrateFallback,
 > = Awaited<
-  [IsDefined<HydrateFallback>, ClientLoaderHydrate]  extends [true, true] ?
+  [undefined extends HydrateFallback ? false : true, ClientLoaderHydrate]  extends [true, true] ?
     IsDefined<ClientLoaderData> extends true ? ClientLoaderData :
     undefined
   :
   [IsDefined<ClientLoaderData>, IsDefined<ServerLoaderData>] extends [true, true] ? ServerLoaderData | ClientLoaderData :
   IsDefined<ClientLoaderData> extends true ?
-    IsDefined<ClientLoaderHydrate> extends true ? ClientLoaderData :
+    ClientLoaderHydrate extends true ? ClientLoaderData :
```

Comment:
This was a bug that didn't show up until I added tests for returning `undefined`. In short `ClientLoaderHydrate` is a `boolean` not a data type, so it should be checked directly.